File: tests/end-to-end/stack_set/csharp/Stack.cs
using Amazon.CDK;
using Amazon.CDK.AWS.IAM;
using Constructs;
using System.Collections.Generic;

namespace StackSetStack
{
    public class StackSetStackProps : StackProps
    {
        public string ModuleName { get; set; }

        public string RoleName { get; set; }

        public string RolePath { get; set; }

    }

    /// <summary>
    /// Deploy required components for StackSet custom resources in this region.  Lambda ARN is exported as StackSetCustomResource
    /// </summary>
    public class StackSetStack : Stack
    {
        public StackSetStack(Construct scope, string id, StackSetStackProps props = null) : base(scope, id, props)
        {
            // Applying default props
            props ??= new StackSetStackProps();
            props.ModuleName ??= "lambda_function";
            props.RoleName ??= "";
            props.RolePath ??= "";


            // Conditions
            bool useRoleName = !(props.RoleName == "");
            bool useRolePath = !(props.RolePath == "");

            // Resources
            var stackSetResourceRole = new CfnRole(this, "StackSetResourceRole", new CfnRoleProps
            {
                RoleName = useRoleName ? props.RoleName : null,
                Path = useRolePath ? props.RolePath : "/",
                AssumeRolePolicyDocument = new Dictionary<string, object>
                {
                    { "Version", "2012-10-17"},
                    { "Statement", new []
                    {
                        new Dictionary<string, object>
                        {
                            { "Effect", "Allow"},
                            { "Principal", new Dictionary<string, object>
                            {
                                { "Service", "lambda.amazonaws.com"},
                            }},
                            { "Action", "sts:AssumeRole"},
                        },
                    }},
                },
                Policies = new []
                {
                    new CfnRole.PolicyProperty
                    {
                        PolicyName = "IAMPassRolePermissions",
                        PolicyDocument = new Dictionary<string, object>
                        {
                            { "Version", "2012-10-17"},
                            { "Statement", new []
                            {
                                new Dictionary<string, object>
                                {
                                    { "Effect", "Allow"},
                                    { "Action", "iam:PassRole"},
                                    { "Resource", "*"},
                                },
                            }},
                        },
                    },
                    new CfnRole.PolicyProperty
                    {
                        PolicyName = "CloudFormationPermissions",
                        PolicyDocument = new Dictionary<string, object>
                        {
                            { "Version", "2012-10-17"},
                            { "Statement", new []
                            {
                                new Dictionary<string, object>
                                {
                                    { "Effect", "Allow"},
                                    { "Action", "cloudformation:*"},
                                    { "Resource", "*"},
                                },
                            }},
                        },
                    },
                    new CfnRole.PolicyProperty
                    {
                        PolicyName = "LambdaPermissions",
                        PolicyDocument = new Dictionary<string, object>
                        {
                            { "Version", "2012-10-17"},
                            { "Statement", new []
                            {
                                new Dictionary<string, object>
                                {
                                    { "Effect", "Allow"},
                                    { "Action", "logs:CreateLogGroup"},
                                    { "Resource", new []
                                    {
                                        $"arn:aws:logs:{Region}:{Account}:*",
                                    }},
                                },
                                new Dictionary<string, object>
                                {
                                    { "Effect", "Allow"},
                                    { "Action", new []
                                    {
                                        "logs:CreateLogStream",
                                        "logs:PutLogEvents",
                                    }},
                                    { "Resource", new []
                                    {
                                        $"arn:aws:logs:{Region}:{Account}:log-group:/aws/lambda/*",
                                    }},
                                },
                            }},
                        },
                    },
                    new CfnRole.PolicyProperty
                    {
                        PolicyName = "S3Permissions",
                        PolicyDocument = new Dictionary<string, object>
                        {
                            { "Version", "2012-10-17"},
                            { "Statement", new []
                            {
                                new Dictionary<string, object>
                                {
                                    { "Effect", "Allow"},
                                    { "Action", new []
                                    {
                                        "s3:Get*",
                                        "s3:List*",
                                    }},
                                    { "Resource", "*"},
                                },
                            }},
                        },
                    },
                },
            });
        }
    }
}

File: tests/end-to-end/stack_set/typescript/stack.ts
import * as cdk from 'aws-cdk-lib';
import * as iam from 'aws-cdk-lib/aws-iam';

export interface StackSetStackProps extends cdk.StackProps {
  /**
   * @default 'lambda_function'
   */
  readonly moduleName?: string;
  /**
   * @default ''
   */
  readonly roleName?: string;
  /**
   * @default ''
   */
  readonly rolePath?: string;
}

/**
 * Deploy required components for StackSet custom resources in this region.  Lambda ARN is exported as StackSetCustomResource
 */
export class StackSetStack extends cdk.Stack {
  public constructor(scope: cdk.App, id: string, props: StackSetStackProps = {}) {
    super(scope, id, props);

    // Applying default props
    props = {
      ...props,
      moduleName: props.moduleName ?? 'lambda_function',
      roleName: props.roleName ?? '',
      rolePath: props.rolePath ?? '',
    };

    // Conditions
    const useRoleName = !(props.roleName! === '');
    const useRolePath = !(props.rolePath! === '');

    // Resources
    const stackSetResourceRole = new iam.CfnRole(this, 'StackSetResourceRole', {
      roleName: useRoleName ? props.roleName! : undefined,
      path: useRolePath ? props.rolePath! : '/',
      assumeRolePolicyDocument: {
        Version: '2012-10-17',
        Statement: [
          {
            Effect: 'Allow',
            Principal: {
              Service: 'lambda.amazonaws.com',
            },
            Action: 'sts:AssumeRole',
          },
        ],
      },
      policies: [
        {
          policyName: 'IAMPassRolePermissions',
          policyDocument: {
            Version: '2012-10-17',
            Statement: [
              {
                Effect: 'Allow',
                Action: 'iam:PassRole',
                Resource: '*',
              },
            ],
          },
        },
        {
          policyName: 'CloudFormationPermissions',
          policyDocument: {
            Version: '2012-10-17',
            Statement: [
              {
                Effect: 'Allow',
                Action: 'cloudformation:*',
                Resource: '*',
              },
            ],
          },
        },
        {
          policyName: 'LambdaPermissions',
          policyDocument: {
            Version: '2012-10-17',
            Statement: [
              {
                Effect: 'Allow',
                Action: 'logs:CreateLogGroup',
                Resource: [
                  `arn:aws:logs:${this.region}:${this.account}:*`,
                ],
              },
              {
                Effect: 'Allow',
                Action: [
                  'logs:CreateLogStream',
                  'logs:PutLogEvents',
                ],
                Resource: [
                  `arn:aws:logs:${this.region}:${this.account}:log-group:/aws/lambda/*`,
                ],
              },
            ],
          },
        },
        {
          policyName: 'S3Permissions',
          policyDocument: {
            Version: '2012-10-17',
            Statement: [
              {
                Effect: 'Allow',
                Action: [
                  's3:Get*',
                  's3:List*',
                ],
                Resource: '*',
              },
            ],
          },
        },
      ],
    });
  }
}
